src/main/scala/org/broadinstitute/dsde/agora/server/business/PermissionBusiness.scala

```diff
@@ -1,72 +1,114 @@
     AgoraEntityPermissionsClient.listEntityPermissions(entity)
   }
 
-  def insertEntityPermission(entity: AgoraEntity, requester: String, accessObject: AccessControl) = {
-    authorizeEntityRequester(entity, requester)
+  def insertEntityPermission(entity: AgoraEntity, requester: String, accessObject: AccessControl): Int = {
+    // Inserts can result in an edit in the DAO layer
+    val entityACLs = getEntityACLs(entity, requester)
+    checkSameRequesterAndPermissions(entityACLs.find(acl => acl.user.equals(requester)), accessObject)
+    authorizeEntityRequester(entityACLs, entity, requester)
     AgoraEntityPermissionsClient.insertEntityPermission(entity, accessObject)
   }
 
-  def batchEntityPermission(entity: AgoraEntity, requester: String, accessObjectList: List[AccessControl]) = {
+  def batchEntityPermission(entity: AgoraEntity, requester: String, accessObjectList: List[AccessControl]): Int = {
     accessObjectList.map(accessObject => insertEntityPermission(entity, requester, accessObject)).sum
   }
 
-  def editEntityPermission(entity: AgoraEntity, requester: String, accessObject: AccessControl) = {
-    authorizeEntityRequester(entity, requester)
+  def editEntityPermission(entity: AgoraEntity, requester: String, accessObject: AccessControl): Int = {
+    val entityACLs = getEntityACLs(entity, requester)
+    checkSameRequesterAndPermissions(entityACLs.find(acl => acl.user.equals(requester)), accessObject)
+    authorizeEntityRequester(entityACLs, entity, requester)
     AgoraEntityPermissionsClient.editEntityPermission(entity, accessObject)
 
   }
 
-  def deleteEntityPermission(entity: AgoraEntity, requester: String, userToRemove: String) = {
-    authorizeEntityRequester(entity, requester)
+  def deleteEntityPermission(entity: AgoraEntity, requester: String, userToRemove: String): Int = {
+    checkSameRequester(requester, userToRemove)
+    authorizeEntityRequester(getEntityACLs(entity, requester), entity, requester)
     AgoraEntityPermissionsClient.deleteEntityPermission(entity, userToRemove)
   }
 
-  def authorizeNamespaceRequester(entity: AgoraEntity, requester: String) = {
-    if (!NamespacePermissionsClient.getNamespacePermission(entity, requester).canManage &&
-        !NamespacePermissionsClient.getNamespacePermission(entity, "public").canManage)
-      throw new NamespaceAuthorizationException(AgoraPermissions(Manage), entity, requester)
+
+  private def authorizeNamespaceRequester(acls: Seq[AccessControl], entity: AgoraEntity, requester: String): Unit = {
+    if (!acls.exists(_.roles.canManage))
+      throw NamespaceAuthorizationException(AgoraPermissions(Manage), entity, requester)
+  }
+
+  private def authorizeEntityRequester(acls: Seq[AccessControl], entity: AgoraEntity, requester: String): Unit = {
+    if (!acls.exists(_.roles.canManage))
+      throw AgoraEntityAuthorizationException(AgoraPermissions(Manage), entity, requester)
+  }
```

Comment:
The problem with that is we then end up making extra db calls in the other calls. The goal in this was to first get the ACLs so logic could be performed on them in two different ways. Moving the getACL calls into the authorize methods means we'll have to call them twice in some places.